runtime/src/accounts.rs

```diff
@@ -1289,7 +1289,9 @@ impl Accounts {
                     );
 
                     if execution_status.is_ok() || is_nonce_account || is_fee_payer {
-                        if account.rent_epoch() == INITIAL_RENT_EPOCH {
+                        if !preserve_rent_epoch_for_rent_exempt_accounts
+                            && account.rent_epoch() == INITIAL_RENT_EPOCH
```

Comment:
When do we still need to take this branch?

We no longer create rent-paying accounts; so in what scenario do we still need to take this branch and collect rent on a new account?

Reply:
This branch is still taken when an account with rent epoch 0 is referenced in a transaction.  On every transaction, there's a rent collection that takes place for new accounts, and we check if an account is new based on `rent_epoch == 0`.

However, if an account has been closed (emptied of lamports), we still try to do this rent collection afterwards, and since that account now owes rent, the runtime reinitializes the account to `Account:;default()`.  This is mostly OK, but for stake accounts, this means that the cache falls out of sync, since the cache update happens after the reinitialization.

Reply:
No reason that I can see, it was simply just never removed.  Do you prefer just removing this whole branch rather than tying it to your feature?

Reply:
yeah, _if that is the case than the branch is never taken anyways_, then lets remove it entirely.
cc @CriesofCarrots to confirm that the branch is redundant.

Reply:
Checking! It might be safest to feature-gate the removal, but depending on the order of operations here, we can maybe convince ourselves that's not necessary

Reply:
Okay, I have confirmed that when we reach this line, all transactions that leave any account in a newly rent-paying state will:
a. have `execution_status.is_err()` (`TransactionError::InvalidRentPayingAccount` or `TransactionError::InsufficientFundsForRent`)
b. not feature a rent-paying nonce account, because SystemProgram enforces this
c. not feature a rent-paying fee_payer account, because such a transaction will have result `TransactionExecutionResult::NotExecuted` and not reach this line

Thus any tx with a rent-paying account will not follow this branch.

Transactions that create rent-exempt accounts still take the branch, and call `collect_from_created_account` which includes https://github.com/solana-labs/solana/blob/e74ad90cdb574c8b8598418a4591804e8cf61b33/runtime/src/rent_collector.rs#L203
So I think the feature gating in this PR is still necessary in order to ensure consistency of the rent_epoch field on account creation when nodes update to the change set.

Handling of RentPaying cases (ie calling `collect_from_existing_account`) should not be necessary at all, though.

Reply:
Thanks so much for looking into all this! And great catch on setting the rent epoch.  @behzadnouri are you ok with using your existing feature flag? Since we're no longer advancing rent epoch with your change, it makes sense to also not set rent epoch at creation with the same feature, but I can be convinced otherwise.

Reply:
> Since we're no longer advancing rent epoch with your change, it makes sense to also not set rent epoch at creation with the same feature, but I can be convinced otherwise.

This sounds correct to me.

Reply:
Then we can remove the block entirely on feature cleanup :)

Reply:
@joncinque yeah, sounds good.